{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining cell type by their molecular features\n",
    "### Single-cell transcriptomics for neuron classification\n",
    "In the previous tutorial we have looked at how electrophysiological and morphological features can be used to define cell types in the brain. However, given that physiological properties can take many different forms under different conditions, it is challenging to scale up the electrophysiological approach as a primary way to define cell types. On the other hand,  molecular approaches are more suitable for the task. \n",
    "\n",
    "One of the molecular approaches widely used for generating cell type taxonomies task is single-cell transcriptomics (scRNA-seq), which involves profiling of the RNA transcripts within each cell, and classifying cells together if they share similar expression of RNA transcripts. \n",
    "\n",
    "A variety of single-cell transcriptomic studies on different regions of the mouse brain have been published, from the Allen Institute for Brain Science:\n",
    "- [Adult mouse cortical cell taxonomy revealed by single cell transcriptomics (Tasic et al. 2016)](https://www.nature.com/articles/nn.4216)\n",
    "- [Shared and distinct transcriptomic cell types across neocortical areas (Tasic et al. 2018)](http://www.nature.com/articles/s41586-018-0654-5)\n",
    "- [A taxonomy of transcriptomic cell types across the isocortex and hippocampal formation (Yao et al. 2021)](https://linkinghub.elsevier.com/retrieve/pii/S0092867421005018)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Patch-seq allows for the collection of transcriptomic and electrophysiological features for the same cells\n",
    "Patch-seq is an experimental technique that allows for the collection of transcriptomic and electrophysiological features simultaneously. After patch-clamp recording are obtained, the cellular content of the cell can be extracted and fragments of mRNA can be sequenced and quantified. This technique enables one to directly relate the transcriptomic features of a given neuron to other identifying characteristics of the same neuron.\n",
    "\n",
    "For the purpose of this tutorial, we will use the Patch-deq dataset generated from mouse primary visual cortex, the methodologies for data generation are detailed in this paper: [Integrated Morphoelectric and Transcriptomic Classification of Cortical GABAergic Cells](https://linkinghub.elsevier.com/retrieve/pii/S009286742031254X)."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tutorial steps\n",
    "1. Import common packages\n",
    "2. Step 2. Obtaining count matrix for the Patch-seq dataset\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1. Import common packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You have installed these packages in the previous tutorial\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import pearsonr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These packages are part of the Python Standard Library, you don't need to install them necause they come with Python\n",
    "import urllib.request\n",
    "import tarfile\n",
    "import io\n",
    "import json"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2. Obtaining count matrix for the Patch-seq dataset\n",
    "### Single-cell transcriptomic data are usually presented in the form of count matrices\n",
    "One of the advantages of using scRNA-seq data for neuron classification is that the data are usually presented as matrices, known as a count matrix, where each row represents one gene and each column represents one cell. Therefore, each entry in the count matrix would store the expression level of one gene in one cell. \n",
    "\n",
    "An example of a count matrix is the [Transcriptomic Explorer](https://celltypes.brain-map.org/rnaseq/mouse_ctx-hpf_10x?selectedVisualization=Heatmap&colorByFeature=Cell+Type&colorByFeatureValue=Gad1), which is an interactive web interface that provides you with access to the transcriptomic datasets generated by Allen Institute for Brain Science. You can select the dataset that you want to look at. \n",
    "\n",
    "However, this web interface only provides you with access to the count matrices generated from snRNA-seq datasets. For the Patch-seq dataset, we can download the count matrix in the form of a comma-separated values (csv) file. \n",
    "### Download count matrix for Patch-seq dataset from NeMO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the code cell below to download the count matrix from [The Neuroscience Multi-omic Archive (NeMO)](https://nemoarchive.org/). This step might take a few minutes, depending on your internet connectivity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = \\\n",
    "\"https://data.nemoarchive.org/other/AIBS/AIBS_patchseq/transcriptome/scell/SMARTseq/processed/analysis/20200611/20200513_Mouse_PatchSeq_Release_count.v2.csv.tar\"\n",
    "response = urllib.request.urlopen(url)\n",
    "tar_file = io.BytesIO(response.read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the count matrix was deposited on NeMO as a tar file, we need to extract the csv file from the tar file using the `tarfile` package available as a part of the Python Standard Library, and then read the csv file into a Pandas DataFrame object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with tarfile.open(fileobj=tar_file) as tar:\n",
    "    csv_file = tar.extractfile(tar.getmembers()[0])\n",
    "count_matrix = pd.read_csv(io.StringIO(csv_file.read().decode('utf-8')))\n",
    "count_matrix.set_index(\"Unnamed: 0\", inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PS0810_E1-50_S88</th>\n",
       "      <th>PS0817_E1-50_S19</th>\n",
       "      <th>PS0817_E1-50_S25</th>\n",
       "      <th>PS0817_E1-50_S26</th>\n",
       "      <th>PS0817_E1-50_S27</th>\n",
       "      <th>PS0817_E1-50_S28</th>\n",
       "      <th>PS0817_E1-50_S46</th>\n",
       "      <th>PS0817_E1-50_S52</th>\n",
       "      <th>PS0830_E1-50_S17</th>\n",
       "      <th>PS0830_E1-50_S19</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-J39ZH_S561_E1-50</th>\n",
       "      <th>SM-J39ZH_S562_E1-50</th>\n",
       "      <th>SM-J39ZH_S565_E1-50</th>\n",
       "      <th>SM-J39ZH_S566_E1-50</th>\n",
       "      <th>SM-J39ZH_S571_E1-50</th>\n",
       "      <th>SM-J39ZH_S576_E1-50</th>\n",
       "      <th>SM-J3A1L_S592_E1-50</th>\n",
       "      <th>SM-J3A1L_S593_E1-50</th>\n",
       "      <th>SM-J3A1L_S603_E1-50</th>\n",
       "      <th>SM-J3A1L_S604_E1-50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>51</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>483</td>\n",
       "      <td>89</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>38</td>\n",
       "      <td>4</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4435 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               PS0810_E1-50_S88  PS0817_E1-50_S19  PS0817_E1-50_S25  \\\n",
       "Unnamed: 0                                                            \n",
       "0610005C13Rik                 0                 0                 0   \n",
       "0610006L08Rik                 0                 0                 0   \n",
       "0610007P14Rik                 0                 0                 0   \n",
       "0610009B22Rik                 0                 0                41   \n",
       "0610009E02Rik                 0                 0                 0   \n",
       "\n",
       "               PS0817_E1-50_S26  PS0817_E1-50_S27  PS0817_E1-50_S28  \\\n",
       "Unnamed: 0                                                            \n",
       "0610005C13Rik                 0                 0                 0   \n",
       "0610006L08Rik                 0                 0                 0   \n",
       "0610007P14Rik                81                51                35   \n",
       "0610009B22Rik                 0                 0                 0   \n",
       "0610009E02Rik                 0                 0                 1   \n",
       "\n",
       "               PS0817_E1-50_S46  PS0817_E1-50_S52  PS0830_E1-50_S17  \\\n",
       "Unnamed: 0                                                            \n",
       "0610005C13Rik                 0                 0                 0   \n",
       "0610006L08Rik                 0                 0                 0   \n",
       "0610007P14Rik                 0                 0               483   \n",
       "0610009B22Rik                 0                25                 0   \n",
       "0610009E02Rik                11                 0                 0   \n",
       "\n",
       "               PS0830_E1-50_S19  ...  SM-J39ZH_S561_E1-50  \\\n",
       "Unnamed: 0                       ...                        \n",
       "0610005C13Rik                 0  ...                    0   \n",
       "0610006L08Rik                 0  ...                    0   \n",
       "0610007P14Rik                89  ...                    0   \n",
       "0610009B22Rik                64  ...                   21   \n",
       "0610009E02Rik                 0  ...                    0   \n",
       "\n",
       "               SM-J39ZH_S562_E1-50  SM-J39ZH_S565_E1-50  SM-J39ZH_S566_E1-50  \\\n",
       "Unnamed: 0                                                                     \n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                    0                   14                    0   \n",
       "0610009B22Rik                    0                   12                   13   \n",
       "0610009E02Rik                    0                    9                    0   \n",
       "\n",
       "               SM-J39ZH_S571_E1-50  SM-J39ZH_S576_E1-50  SM-J3A1L_S592_E1-50  \\\n",
       "Unnamed: 0                                                                     \n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                   14                    0                   20   \n",
       "0610009B22Rik                   23                   38                    0   \n",
       "0610009E02Rik                    1                    0                    0   \n",
       "\n",
       "               SM-J3A1L_S593_E1-50  SM-J3A1L_S603_E1-50  SM-J3A1L_S604_E1-50  \n",
       "Unnamed: 0                                                                    \n",
       "0610005C13Rik                    0                    0                    0  \n",
       "0610006L08Rik                    0                    0                    0  \n",
       "0610007P14Rik                   38                    4                   56  \n",
       "0610009B22Rik                    0                    0                   54  \n",
       "0610009E02Rik                   13                    0                   50  \n",
       "\n",
       "[5 rows x 4435 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the first 5 rows of the count matrix\n",
    "count_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4435 cells and 45768 genes in the count matrix.\n"
     ]
    }
   ],
   "source": [
    "print(f\"There are {count_matrix.shape[1]} cells and {count_matrix.shape[0]} genes in the count matrix.\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subsetting for genes that code for voltage-gated ion channels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/Users/xunuo/projects/leafcutter/data/IC_list.json\", \"r\") as f:\n",
    "    IC_list = json.load(f)\n",
    "count_matrix = count_matrix.loc[count_matrix.index.isin(IC_list[\"VGIC\"])]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove genes that are not expressed in any cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_matrix = count_matrix[count_matrix.sum(axis = 1) != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_matrix = count_matrix.dropna(axis = 0, how=\"any\").dropna(axis = 1, how=\"any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PS0810_E1-50_S88</th>\n",
       "      <th>PS0817_E1-50_S19</th>\n",
       "      <th>PS0817_E1-50_S25</th>\n",
       "      <th>PS0817_E1-50_S26</th>\n",
       "      <th>PS0817_E1-50_S27</th>\n",
       "      <th>PS0817_E1-50_S28</th>\n",
       "      <th>PS0817_E1-50_S46</th>\n",
       "      <th>PS0817_E1-50_S52</th>\n",
       "      <th>PS0830_E1-50_S17</th>\n",
       "      <th>PS0830_E1-50_S19</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-J39ZH_S561_E1-50</th>\n",
       "      <th>SM-J39ZH_S562_E1-50</th>\n",
       "      <th>SM-J39ZH_S565_E1-50</th>\n",
       "      <th>SM-J39ZH_S566_E1-50</th>\n",
       "      <th>SM-J39ZH_S571_E1-50</th>\n",
       "      <th>SM-J39ZH_S576_E1-50</th>\n",
       "      <th>SM-J3A1L_S592_E1-50</th>\n",
       "      <th>SM-J3A1L_S593_E1-50</th>\n",
       "      <th>SM-J3A1L_S603_E1-50</th>\n",
       "      <th>SM-J3A1L_S604_E1-50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cacna1a</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cacna1b</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cacna1c</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cacna1d</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cacna1e</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trpv2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trpv3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trpv4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trpv5</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trpv6</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows × 4354 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PS0810_E1-50_S88  PS0817_E1-50_S19  PS0817_E1-50_S25  \\\n",
       "Unnamed: 0                                                         \n",
       "Cacna1a                False             False             False   \n",
       "Cacna1b                False             False             False   \n",
       "Cacna1c                False             False             False   \n",
       "Cacna1d                False             False             False   \n",
       "Cacna1e                False             False             False   \n",
       "...                      ...               ...               ...   \n",
       "Trpv2                  False             False             False   \n",
       "Trpv3                  False             False             False   \n",
       "Trpv4                  False             False             False   \n",
       "Trpv5                  False             False             False   \n",
       "Trpv6                  False             False             False   \n",
       "\n",
       "            PS0817_E1-50_S26  PS0817_E1-50_S27  PS0817_E1-50_S28  \\\n",
       "Unnamed: 0                                                         \n",
       "Cacna1a                False             False             False   \n",
       "Cacna1b                False             False             False   \n",
       "Cacna1c                False             False             False   \n",
       "Cacna1d                False             False             False   \n",
       "Cacna1e                False             False             False   \n",
       "...                      ...               ...               ...   \n",
       "Trpv2                  False             False             False   \n",
       "Trpv3                  False             False             False   \n",
       "Trpv4                  False             False             False   \n",
       "Trpv5                  False             False             False   \n",
       "Trpv6                  False             False             False   \n",
       "\n",
       "            PS0817_E1-50_S46  PS0817_E1-50_S52  PS0830_E1-50_S17  \\\n",
       "Unnamed: 0                                                         \n",
       "Cacna1a                False             False             False   \n",
       "Cacna1b                False             False             False   \n",
       "Cacna1c                False             False             False   \n",
       "Cacna1d                False             False             False   \n",
       "Cacna1e                False             False             False   \n",
       "...                      ...               ...               ...   \n",
       "Trpv2                  False             False             False   \n",
       "Trpv3                  False             False             False   \n",
       "Trpv4                  False             False             False   \n",
       "Trpv5                  False             False             False   \n",
       "Trpv6                  False             False             False   \n",
       "\n",
       "            PS0830_E1-50_S19  ...  SM-J39ZH_S561_E1-50  SM-J39ZH_S562_E1-50  \\\n",
       "Unnamed: 0                    ...                                             \n",
       "Cacna1a                False  ...                False                False   \n",
       "Cacna1b                False  ...                False                False   \n",
       "Cacna1c                False  ...                False                False   \n",
       "Cacna1d                False  ...                False                False   \n",
       "Cacna1e                False  ...                False                False   \n",
       "...                      ...  ...                  ...                  ...   \n",
       "Trpv2                  False  ...                False                False   \n",
       "Trpv3                  False  ...                False                False   \n",
       "Trpv4                  False  ...                False                False   \n",
       "Trpv5                  False  ...                False                False   \n",
       "Trpv6                  False  ...                False                False   \n",
       "\n",
       "            SM-J39ZH_S565_E1-50  SM-J39ZH_S566_E1-50  SM-J39ZH_S571_E1-50  \\\n",
       "Unnamed: 0                                                                  \n",
       "Cacna1a                   False                False                False   \n",
       "Cacna1b                   False                False                False   \n",
       "Cacna1c                   False                False                False   \n",
       "Cacna1d                   False                False                False   \n",
       "Cacna1e                   False                False                False   \n",
       "...                         ...                  ...                  ...   \n",
       "Trpv2                     False                False                False   \n",
       "Trpv3                     False                False                False   \n",
       "Trpv4                     False                False                False   \n",
       "Trpv5                     False                False                False   \n",
       "Trpv6                     False                False                False   \n",
       "\n",
       "            SM-J39ZH_S576_E1-50  SM-J3A1L_S592_E1-50  SM-J3A1L_S593_E1-50  \\\n",
       "Unnamed: 0                                                                  \n",
       "Cacna1a                   False                False                False   \n",
       "Cacna1b                   False                False                False   \n",
       "Cacna1c                   False                False                False   \n",
       "Cacna1d                   False                False                False   \n",
       "Cacna1e                   False                False                False   \n",
       "...                         ...                  ...                  ...   \n",
       "Trpv2                     False                False                False   \n",
       "Trpv3                     False                False                False   \n",
       "Trpv4                     False                False                False   \n",
       "Trpv5                     False                False                False   \n",
       "Trpv6                     False                False                False   \n",
       "\n",
       "            SM-J3A1L_S603_E1-50  SM-J3A1L_S604_E1-50  \n",
       "Unnamed: 0                                            \n",
       "Cacna1a                   False                False  \n",
       "Cacna1b                   False                False  \n",
       "Cacna1c                   False                False  \n",
       "Cacna1d                   False                False  \n",
       "Cacna1e                   False                False  \n",
       "...                         ...                  ...  \n",
       "Trpv2                     False                False  \n",
       "Trpv3                     False                False  \n",
       "Trpv4                     False                False  \n",
       "Trpv5                     False                False  \n",
       "Trpv6                     False                False  \n",
       "\n",
       "[142 rows x 4354 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.isnan(count_matrix)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove cells that do not express any of the ion channel genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalizing gene counts CPM"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `count_matrix` DataFrame contains the raw read counts. It cannot be used to compare gene expression levels between cells due to the need to account for differences in gene length, total number of reads per samples, and sequencing biases. One common way to normalize gene expression levels is to calculate counts per million (CPM), which is obtained by dividing read counts for genes by the total number of reads in that cell and multiplying the results by a million, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import precomputed library sizes for each cell, each row contains the library size for one cell\n",
    "lib_size = pd.read_csv(\n",
    "    \"/Users/xunuo/projects/ephys-tutorial/transcriptomic_tutorial/data/lib_size.csv\", \n",
    "    sep = \"\\t\", index_col = 0)\n",
    "# Remove cells that contain 0 reads or NA values\n",
    "lib_size = lib_size.dropna() \\\n",
    "    .loc[lib_size.library_size != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the cells that have been removed in the previous step from lib_size\n",
    "count_matrix = count_matrix.loc[:, count_matrix.columns.isin(lib_size.index)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "><b>Task:</b> Calculating counts per million for `count_matrix`, using library sizes provided in `lib_size`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "cpm = count_matrix.apply(lambda x: x/lib_size.library_size * 1e6, axis = 1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3. Constructing Gene-ephys correlation matrix"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In `cpm`, each column is labelled by a `transcriptomics_sample_id`. However, this ID only identifies the transcriptomic sample collected from each cell. To correlate transcriptomic data with electrophysiological data, we need to know which cell each transcriptomic sample corresponds to. To achieve this, we first need to convert the `transcriptomics_sample_id` label to `cell_specimen_id`, which identifies one cell uniquely. Then, we can merge transcriptomic and electrophysiological data based on this shared label:  `cell_specimen_id`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From the metadata file downloaded from the Allen Institute website, we can a mapping from transcriptomics_sample_id to cell_specimen_id\n",
    "metadata = \\\n",
    "    pd.read_csv(\"/Users/xunuo/projects/leafcutter/data/20200625_patchseq_metadata_mouse.csv\", usecols = [\"cell_specimen_id\",  \"transcriptomics_sample_id\"])\\\n",
    "    .set_index(\"transcriptomics_sample_id\")\\\n",
    "    .to_dict()[\"cell_specimen_id\"]\n",
    "# Convert the column names of cpm from transcriptomics_sample_id to cell_specimen_id\n",
    "cpm = cpm.rename(columns = metadata)    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import extracted electrophysiological features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_list = ['Specimen ID', 'Avg ISI', 'Fast trough V (long square) (millivolts)', 'Trough V (long square) (millivolts)', 'Peak V (long square) (millivolts)', 'Peak V (short square) (millivolts)',\\\n",
    "                   'Upstroke/downstroke ratio (long square)', 'Threshold V (long square) (millivolts)', 'F I curve slope', 'sag', 'tau', 'Vrest (millivolts)']\n",
    "ephys_data = pd.read_csv(\n",
    "    \"/Users/xunuo/projects/ephys-tutorial/transcriptomic_tutorial/data/SpecimenMetadata.csv\", \n",
    "    usecols=column_list, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop columns that contain any NA\n",
    "ephys_data = ephys_data.dropna(axis = 1, how = \"any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ephys_data = (ephys_data - ephys_data.apply(np.mean)) / ephys_data.apply(np.std)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_matrix = np.corrcoef(np.hstack([cpm.T.reindex(ephys_data.index).to_numpy(), ephys_data]), rowvar=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/xunuo/miniconda3/lib/python3.11/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[:, None]\n",
      "/Users/xunuo/miniconda3/lib/python3.11/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide\n",
      "  c /= stddev[None, :]\n"
     ]
    }
   ],
   "source": [
    "a = np.corrcoef(cpm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "601506507     0.0\n",
       "601790961     0.0\n",
       "601803754     0.0\n",
       "601808698     0.0\n",
       "601810307     0.0\n",
       "             ... \n",
       "992825872     0.0\n",
       "1003305641    0.0\n",
       "1003313398    0.0\n",
       "1003321825    0.0\n",
       "1003336938    0.0\n",
       "Name: Trpc2, Length: 4354, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm.iloc[123]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_list = ['Specimen ID', 'Avg ISI', 'Fast trough V (long square) (millivolts)', 'Peak V (long square) (millivolts)', 'Peak V (short square) (millivolts)',\\\n",
    "                   'Upstroke/downstroke ratio (long square)', 'Threshold V (long square) (millivolts)', 'F I curve slope']\n",
    "ephys_data = pd.read_csv(\n",
    "    \"/Users/xunuo/projects/ephys-tutorial/transcriptomic_tutorial/data/SpecimenMetadata.csv\", \n",
    "    usecols=column_list, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Cacna1a</th>\n",
       "      <th>Cacna1b</th>\n",
       "      <th>Cacna1c</th>\n",
       "      <th>Cacna1d</th>\n",
       "      <th>Cacna1e</th>\n",
       "      <th>Cacna1f</th>\n",
       "      <th>Cacna1g</th>\n",
       "      <th>Cacna1h</th>\n",
       "      <th>Cacna1i</th>\n",
       "      <th>Cacna1s</th>\n",
       "      <th>...</th>\n",
       "      <th>Trpm5</th>\n",
       "      <th>Trpm6</th>\n",
       "      <th>Trpm7</th>\n",
       "      <th>Trpm8</th>\n",
       "      <th>Trpv1</th>\n",
       "      <th>Trpv2</th>\n",
       "      <th>Trpv3</th>\n",
       "      <th>Trpv4</th>\n",
       "      <th>Trpv5</th>\n",
       "      <th>Trpv6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Specimen ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>888317699</th>\n",
       "      <td>282.559751</td>\n",
       "      <td>121.365425</td>\n",
       "      <td>209.289603</td>\n",
       "      <td>103.705440</td>\n",
       "      <td>228.828309</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.544661</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.751489</td>\n",
       "      <td>69.512705</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18.411473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765853817</th>\n",
       "      <td>270.433282</td>\n",
       "      <td>53.138955</td>\n",
       "      <td>75.477624</td>\n",
       "      <td>115.416457</td>\n",
       "      <td>13.200123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>25.046387</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36.554186</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671876425</th>\n",
       "      <td>140.326381</td>\n",
       "      <td>31.485975</td>\n",
       "      <td>175.699513</td>\n",
       "      <td>91.348198</td>\n",
       "      <td>2.332294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.388716</td>\n",
       "      <td>0.388716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.435787</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.329178</td>\n",
       "      <td>20.990650</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906699622</th>\n",
       "      <td>67.601863</td>\n",
       "      <td>27.336272</td>\n",
       "      <td>57.627817</td>\n",
       "      <td>44.698499</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.171727</td>\n",
       "      <td>8.865818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>50.978454</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.369409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778301240</th>\n",
       "      <td>259.567011</td>\n",
       "      <td>116.411871</td>\n",
       "      <td>53.486536</td>\n",
       "      <td>132.143205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>56.632802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701163677</th>\n",
       "      <td>4.272623</td>\n",
       "      <td>195.930266</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>94.608072</td>\n",
       "      <td>39.063978</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>23.194237</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.155620</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671529949</th>\n",
       "      <td>116.565670</td>\n",
       "      <td>30.548245</td>\n",
       "      <td>131.035891</td>\n",
       "      <td>142.692458</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.901430</td>\n",
       "      <td>3.215605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.470221</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.401951</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.803901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766764831</th>\n",
       "      <td>151.448278</td>\n",
       "      <td>104.715667</td>\n",
       "      <td>100.821282</td>\n",
       "      <td>88.272711</td>\n",
       "      <td>182.170643</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.057931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.923350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673132928</th>\n",
       "      <td>259.026922</td>\n",
       "      <td>89.995251</td>\n",
       "      <td>323.119481</td>\n",
       "      <td>61.103787</td>\n",
       "      <td>0.332086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.227293</td>\n",
       "      <td>0.332086</td>\n",
       "      <td>3.652944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.156183</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.275947</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.649201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978205949</th>\n",
       "      <td>42.578979</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.990209</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4284 rows × 143 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0      Cacna1a     Cacna1b     Cacna1c     Cacna1d     Cacna1e  \\\n",
       "Specimen ID                                                               \n",
       "888317699    282.559751  121.365425  209.289603  103.705440  228.828309   \n",
       "765853817    270.433282   53.138955   75.477624  115.416457   13.200123   \n",
       "671876425    140.326381   31.485975  175.699513   91.348198    2.332294   \n",
       "906699622     67.601863   27.336272   57.627817   44.698499    0.000000   \n",
       "778301240    259.567011  116.411871   53.486536  132.143205    0.000000   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "701163677      4.272623  195.930266    0.000000   94.608072   39.063978   \n",
       "671529949    116.565670   30.548245  131.035891  142.692458    0.000000   \n",
       "766764831    151.448278  104.715667  100.821282   88.272711  182.170643   \n",
       "673132928    259.026922   89.995251  323.119481   61.103787    0.332086   \n",
       "978205949     42.578979    0.000000    0.000000    0.990209    0.000000   \n",
       "\n",
       "Unnamed: 0    Cacna1f    Cacna1g   Cacna1h    Cacna1i  Cacna1s  ...  Trpm5  \\\n",
       "Specimen ID                                                     ...          \n",
       "888317699    0.000000  22.544661  0.000000   0.000000      0.0  ...    0.0   \n",
       "765853817    0.000000   0.000000  0.000000  25.046387      0.0  ...    0.0   \n",
       "671876425    0.000000   0.388716  0.388716   0.000000      0.0  ...    0.0   \n",
       "906699622    5.171727   8.865818  0.000000   0.000000      0.0  ...    0.0   \n",
       "778301240    0.000000   0.000000  0.000000   0.000000      0.0  ...    0.0   \n",
       "...               ...        ...       ...        ...      ...  ...    ...   \n",
       "701163677    0.000000  23.194237  0.000000   0.000000      0.0  ...    0.0   \n",
       "671529949    0.000000  20.901430  3.215605   0.000000      0.0  ...    0.0   \n",
       "766764831    0.000000   6.057931  0.000000   0.000000      0.0  ...    0.0   \n",
       "673132928    0.000000  28.227293  0.332086   3.652944      0.0  ...    0.0   \n",
       "978205949    0.000000   0.000000  0.000000   0.000000      0.0  ...    0.0   \n",
       "\n",
       "Unnamed: 0      Trpm6      Trpm7  Trpm8     Trpv1      Trpv2  Trpv3     Trpv4  \\\n",
       "Specimen ID                                                                     \n",
       "888317699    0.751489  69.512705    0.0  0.000000  18.411473    0.0  0.000000   \n",
       "765853817    0.000000  36.554186    0.0  0.000000   0.000000    0.0  0.000000   \n",
       "671876425    0.000000  19.435787    0.0  9.329178  20.990650    0.0  0.000000   \n",
       "906699622    0.000000  50.978454    0.0  0.000000   0.000000    0.0  0.369409   \n",
       "778301240    0.000000  56.632802    0.0  0.000000   0.000000    0.0  0.000000   \n",
       "...               ...        ...    ...       ...        ...    ...       ...   \n",
       "701163677    0.000000   9.155620    0.0  0.000000   0.000000    0.0  0.000000   \n",
       "671529949    0.000000  14.470221    0.0  0.000000   0.401951    0.0  0.000000   \n",
       "766764831    0.000000   6.923350    0.0  0.000000   0.000000    0.0  0.000000   \n",
       "673132928    0.000000  47.156183    0.0  0.000000  15.275947    0.0  0.000000   \n",
       "978205949    0.000000   0.000000    0.0  0.000000   0.000000    0.0  0.000000   \n",
       "\n",
       "Unnamed: 0   Trpv5     Trpv6  \n",
       "Specimen ID                   \n",
       "888317699      0.0  0.000000  \n",
       "765853817      0.0  0.000000  \n",
       "671876425      0.0  0.000000  \n",
       "906699622      0.0  0.000000  \n",
       "778301240      0.0  0.000000  \n",
       "...            ...       ...  \n",
       "701163677      0.0  0.000000  \n",
       "671529949      0.0  0.803901  \n",
       "766764831      0.0  0.000000  \n",
       "673132928      0.0  4.649201  \n",
       "978205949      0.0  0.000000  \n",
       "\n",
       "[4284 rows x 143 columns]"
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpm.T.reindex(ephys_data.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4284, 7)"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ephys_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
